source/stylist/fortran.py

```diff
@@ -195,3 +196,98 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
                     issues.append(Issue(description.format(module=module)))
 
         return issues
+
+
+class MissingPointerInit(FortranRule):
+    """
+    Catches cases where a pointer is declared without being initialised.
+    """
+
+    def __init__(self, default='null()'):
+        """
+        Constructs a MissingPointerInit rule object taking a default
+        assignment.
+
+        Todo: Obviously the default is not used as we don't support coercing
+              source at the moment.
+
+        @param default: Target to be used if missing assignment found.
+        """
+        self._default = default
+
+    def examine_fortran(self, subject: FortranSource):
+        issues: List[Issue] = []
+
+        candidates: List[Fortran2003.StmtBase] = []
+        # Component variables
+        candidates.extend(
+            subject.find_all(Fortran2003.Data_Component_Def_Stmt))
+        # Component Procedure
+        candidates.extend(
+            subject.find_all(Fortran2003.Proc_Component_Def_Stmt))
+        # Procedure declaration
+        candidates.extend(
+            subject.find_all(Fortran2003.Procedure_Declaration_Stmt))
+        # Variable declaration
+        candidates.extend(
+            subject.find_all(Fortran2003.Type_Declaration_Stmt))
+
+        return_variable = ''
+        for candidate in candidates:
+            if isinstance(candidate,  # Is variable
+                          (Fortran2003.Data_Component_Def_Stmt,
+                           Fortran2003.Type_Declaration_Stmt)):
+                if isinstance(candidate.parent.parent,
+                              Fortran2003.Function_Subprogram):
+                    # Is contained in a function
+                    function_block: Fortran2003.Function_Subprogram \
+                        = candidate.parent.parent
+                    statement = None
+                    for thing in function_block.children:
+                        if isinstance(thing, Fortran2003.Function_Stmt):
+                            statement = thing
+                            break
+                    if statement is None:
+                        message = "Malformed parse tree: Function subprogram" \
+                                  " without Function statement"
+                        raise Exception(message)
+                    suffix = statement.items[3]
+                    if isinstance(suffix, Fortran2003.Suffix):
+                        if isinstance(suffix.items[0], Fortran2003.Name):
+                            # Is return variable
+                            return_variable = str(suffix.items[0])
+
+            problem = 'Declaration of pointer "{0}" without initialisation.'
```

Comment:
Would it be possible to add a line number output to help find the problem pointer in a file?

Reply:
Obviously line numbers would be great but I avoided it previously because it looked like a lot of work. Turns out to be easier than I feared. The resulst wont be entirely acurate since fparser has to deal with the fact that a statement can be split across lines so provides a span. I just took the first line of the span.